Improve DHT publishing

fiaxh opened the pull request:
Previously, the own DHT entry was published on every new epoch. That means
- It might try to publish before the DHT finished setting up
- It republishes the DHT although we configured the library to automatically do that every 60 seconds.

This PR adds an event to the Network that is fired when DHT is done bootstrapping. The validator listens for this event and publishes the own record upon receiving it.

hrxi commented on validator/src/validator.rs:
```diff
@@ -892,6 +897,17 @@ where
+        // Check if DHT is bootstrapped and we can publish our record
+        while let Poll::Ready(Some(result)) = self.network_event_rx.poll_next_unpin(cx) {
+            match result {
+                Ok(NetworkEvent::DhtBootstrapped) => {
```
AFAICT, the event isn't sent anywhere.

hrxi commented on validator/src/validator.rs:
```diff
@@ -322,19 +338,9 @@ where
-        let key = self.voting_key();
         let network = Arc::clone(&self.network);
 
-        // TODO might better be done without the task.
```
Is the "might better be done without the task." fixed by this commit? If not, it should probably stay.

fiaxh committed:
Improve DHT publishing

Previously, the own DHT entry was published on every new epoch. That means it might try to publish before the DHT finished setting up. Also, it republished the DHT although we configured the library to automatically do that every 60 seconds.
This PR adds an event to the Network that is fired when DHT is done bootstrapping. The validator listens for this event and publishes the own record upon receiving it.

nibhar commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -337,7 +337,7 @@ impl<N: Network> Future for RemoteEventDispatcher<N> {
                     self.state.write().remove_peer(&peer_id);
                 }
-                Ok(NetworkEvent::PeerJoined(_peer_id, _)) => {}
+                Ok(_) => {}
```
I think adding information is better than removing it. Very much personal preference though.

```suggestion
                Ok(NetworkEvent::PeerJoined(_peer_id, _peer_info)) => {}
                Ok(NetworkEvent::DhtBootstrapped => {}
```

hrxi replied:
I think for a "listen for events" it makes sense to ignore all that you don't want to listen for. Otherwise you'll need to update each location when you add another event.

nibhar commented on validator/src/validator.rs:
```diff
@@ -249,8 +251,22 @@ where
         self.control_mempool_monitor.clone()
     }
 
+    pub(crate) fn publish_dht(&self) {
```
Looks to me like the `pub(crate)` might not be necessary.

fiaxh replied:
Indeed

nibhar commented on validator/src/validator.rs:
```diff
@@ -892,6 +898,17 @@ where
+                    self.publish_dht();
+                }
+                Ok(_) => {}
+                Err(_) => break,
```
What does this case mean? Is this safe to just ignore? Will this recover?

fiaxh replied:
That error is thrown by the broadcast stream in the following case:

> The receiver lagged too far behind. Attempting to receive again will return the oldest message still retained by the channel.

Thus, it would be better to continue the loop instead of breaking (although that would have also continued in the next poll). I adjusted the code accordingly.

hrxi reviewed (approved):
LGTM.
